In [1]:
import pandas as pd
import datetime
import random
import calendar

### Using Tabula to import chart data from Bureau of Labor Statistics
---

In [2]:
#import tabula
# file = '../resources/cesan.pdf'
# table = tabula.read_pdf(file, pages=5)

In [3]:
# test = table[0].copy()#.columns({'Unnamed: 0':'first'})

In [4]:
# test.to_csv('consumer_spending.csv', index=False)

In [5]:
table = pd.read_csv('consumer_spending.csv').reset_index(drop=True)
table = table.rename(columns={table.columns[0]:'category',table.columns[1]:'2018', table.columns[2]:'2019', table.columns[3]:'2020'})
# table['category'] = table['category'].str.replace(r'[^\w\s]+','')
# table['2018'] = table['2018'].str.replace(r'[^\w\s]+','')
for i in table.columns:
    table[i] = table[i].str.replace(r'[^\w\s]+','')
table= table[['category', '2018', '2019', '2020']]
table= table.iloc[2:18].reset_index(drop=True)
#table

keep_list = ['Food', 'Housing', 'Apparel and services']
#for i in keep_list:
keep = table.loc[(table.category == 'Food')|
                 (table.category == 'Housing')|
                 (table.category == 'Apparel and services')|
                 (table.category == 'Average income before taxes')|
                 (table.category == 'Average annual expenditures')].reset_index(drop=True)

# 2nd table on page 6 is too messy easier to type out answers 2018
appendlist = [['transportation', 9761, 10742, 9826],
['healthcare', 4968, 5193, 5177],
['entertainment', 3226, 3050, 2912],
['personal_care_products', 768, 768, 864],
['reading', 108, 92, 114],
['education', 1407, 1443, 1271],
['tobacco', 347, 320, 315],
['misc', 993, 899, 907],
['cash_contrib', 1888, 1995, 2283],
['personal_insurance_pensions', 7296, 7165, 7246]]

for i in appendlist:
    #print(i)
    keep2 = pd.Series(i, index=['category', '2018', '2019', '2020'])
    #print(keep2)
    keep  = keep.append(keep2, ignore_index = True)
#lowering categories
keep.category = keep.category.str.lower()
keep

category   2018   2019   2020
0   average income before taxes  78635  82852  84352
1   average annual expenditures  61224  63036  61334
2                          food   7923   8169   7316
3                       housing  20091  20679  21409
4          apparel and services   1866   1883   1434
5                transportation   9761  10742   9826
6                    healthcare   4968   5193   5177
7                 entertainment   3226   3050   2912
8        personal_care_products    768    768    864
9                       reading    108     92    114
10                    education   1407   1443   1271
11                      tobacco    347    320    315
12                         misc    993    899    907
13                 cash_contrib   1888   1995   2283
14  personal_insurance_pensions   7296   7165   7246

In [6]:
# Adding percentage of average income to category column
keep['2018'] = keep['2018'].astype(int)
keep['2019'] = keep['2019'].astype(int)
keep['2020'] = keep['2020'].astype(int)
add_perc = []
for i in range(len(keep['2020'])):
    #print(keep['2018'][i])
    total = keep['2020'].sum()
    add_perc.append(round(keep['2020'][i]/keep['2020'][2:].sum()*100, 2))

keep['percentage_2020'] = add_perc
keep['percentage_2020'][0] = 0
keep['percentage_2020'][1] = 0
add_perc = []
for i in range(len(keep['2019'])):
    #print(keep['2018'][i])
    total = keep['2019'].sum()
    add_perc.append(round(keep['2019'][i]/keep['2019'][2:].sum()*100, 2))

keep['percentage_2019'] = add_perc
keep['percentage_2019'][0] = 0
keep['percentage_2019'][1] = 0
add_perc = []
for i in range(len(keep['2018'])):
    #print(keep['2018'][i])
    total = keep['2018'].sum()
    add_perc.append(round(keep['2018'][i]/keep['2018'][2:].sum()*100, 2))

keep['percentage_2018'] = add_perc
keep['percentage_2018'][0] = 0
keep['percentage_2018'][1] = 0
keep

/Users/Jen/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if sys.path[0] == '':
/Users/Jen/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  del sys.path[0]
/Users/Jen/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/Users/Jen/anaconda3/lib/python3.7/site-

category   2018   2019   2020  percentage_2020  \
0   average income before taxes  78635  82852  84352             0.00   
1   average annual expenditures  61224  63036  61334             0.00   
2                          food   7923   8169   7316            11.98   
3                       housing  20091  20679  21409            35.05   
4          apparel and services   1866   1883   1434             2.35   
5                transportation   9761  10742   9826            16.09   
6                    healthcare   4968   5193   5177             8.48   
7                 entertainment   3226   3050   2912             4.77   
8        personal_care_products    768    768    864             1.41   
9                       reading    108     92    114             0.19   
10                    education   1407   1443   1271             2.08   
11                      tobacco    347    320    315             0.52   
12                         misc    993    899    907             1.49   
13                 cash_contrib   1888   1995   2283             3.74   
14  personal_insurance_pensions   7296   7165   7246            11.86   

    percentage_2019  percentage_2018  
0              0.00             0.00  
1              0.00             0.00  
2             13.09            13.07  
3             33.14            33.13  
4              3.02             3.08  
5             17.22            16.10  
6              8.32             8.19  
7              4.89             5.32  
8              1.23             1.27  
9              0.15             0.18  
10             2.31             2.32  
11             0.51             0.57  
12             1.44             1.64  
13             3.20             3.11  
14            11.48            12.03

# Generating Seed Data
---
- Making selected identifiers and transactions for sample data

In [33]:
transaction_dict = {
    'gas':['QuikTrip 0000 00000000 ATLANTA GA',
       'SHELL OIL 00000000000 EAST POINT GA',
       'EXXON 000000-00 ATLANTA, GA',
        'QuikTrip 1234 00000000 ATLANTA GA',
       'SHELL OIL 1111000000 EAST POINT GA',
       'EXXON 222200-00 ATLANTA, GA'],
    
'groceries' : ['KROGER #000 STONE MOUNTAIN GA',
            'PUBLIX SUPERMAR #0000 DECATOR, GA',
            'WHOLE FOODS 000000 ATLANTA, GA',
            'COSTCO 000 000000000000 DULUTH, GA'],

'grooming' : ['CRAFTED HAIR SALON JUo000 LAWRENCEVILLE GA',
           'BATH AND BODY WORKS 0000 LAWRENCEVILLE GA'],

'homeimprov':['LOWES #00000* BUFORD GA',
             'TRACTOR SUPPLY CO #0000 GRAYSON GA',
             'THE HOME DEPOT #0000 NORCROSS GA'],

'medical' : ['CVS/PHARMACY #0000 SUWANEE GA', 
          'WALGREENS #00000 WINDER GA',
          'DRROBERTPARK OPTOMETRIST & ASSoc SNELLVILLE, GA',
          'GREATER ATLANTA PEDIATRICS STONEMOUNTAIN GA',
          'EMORYUNV HOSPITAL 78979879 ATLANTA GA',
             'ATLANTA GYO&OB *0000000 DECATUR GA '],

'conveyance' :['E ZPASS MD CSC0000000 000-0000000 MD',
              'AUTOZONE 0000 STONEMOUNTAIN GA',
              'ADVANCE AUTO PARTS #0000 LITHONIA GA',
             'EXPRESSOILCHANGE 00*0000000-0 SNELLVILLE, GA',
             'NORTHERN TOOL 00 000000 000 SNELLVILLE, GA'],

'takeaway': ['PANERA BREAD #000000 P ATLANTA GA', 
        'CHICK-FIL-A #00000 TUCKER GA',
        "WENDY'S LOGAN LOGANVILLE GA",
        'BOJANGLES 0000 DULUTH GA',
        'COOK OUT LITHONIA GA 0 LITHONIA GA',
        'TROPICAL SMOOTHIE CAFE GA DECATUR GA',
        'HARDEES 0000 LILBURN GA',
        'PANDA EXPRESS 00000H TUCKER GA',
        'KFC G00000 DECATUR GA',
        'Subway 00000 Lithonia GA',
        'STARBUCKS STORE 0000 ATLANTA GA'],

'utility' :['COX INTERNET 1233456 GA',
           'VZWRLSS*APOCC VISW 000-000-0000 FL',
          'T-MOBILE 0000 00000 00000 ATLANTA GA',
          'GEORGIAPOWERCO 000000 ATLANTA GA',
          'CITYOFDECATUR PUBLIC WORKS 0000000 GA'],

'clothing' : ['ROSS STORES # 0000 TUCKER GA',
           'BATH AND BODY WORKS 0000 STONECREST GA',
           'WAL-MART #0000 STONEMOUNTAIN GA',
           "KOHL'S #0000 SNELLVILLE GA",
           'TARGET 00000000 TUCKER GA',
           'WM SUPERCENTER #0000 CHAMBLEE GA'],

'holiday' : [
    'AMERICAN AIR 00-00000000000 MINNEAPOLIS, MN',
    'EXPEDIA 00000000000000 EXPEDIA.COM WA',
    'UNITED 000000000000 000-000-0000 TX'
    'SANDALSOCHI JAMAICA *008900088899 JAMAICA'],

'streaming' : ['Disney PLUS Burbank CA',
             'AMZN Mktp US*0A0000000 Amzn.com/bill WA',
             'APPLE.COM/BILL 000-000-0000 CA'],

'work': ['PAYPAL *LINKEDIN 000-000-0000 CA',
       'OFFICEMAX/DEPOT 0000 peachtree corners GA',
        'BEST BUY 000000 TUCKER GA'],
  
'gift':['Etsy.com - leatherwork00 000-0000000 NY', 
      'SHUTTERFLY, INC. 000-000000000 CA',
     'PeachTree FLOWERS 0000000 ATLANTA, GA'],

'baby': ['buybuy BABY 000000 Snellville Ga',
       'POTTERYBARN KIDS 000 000000 ATL GA',
       'CARTERS KIDS 00 00 00000 LAWERENCEVILLE GA',
       'BABYGAP 0-0000-000 Atlanta GA'],

'wedding' : ['TIFFANY&CO *90-000009- ATLANTA GA',
          'helzbergDiamonds 70000-000 ATLANTA GA',
          'BrianneMcMullan Events llc *999*9999****** GA',
          'ATLBOTANICALGAR ATLANTA GA',
          'the White Magnolia Bridal ATLANTA GA',
          'RACHMAKESCAKES ATL GA'],

'furniture':['IKEACO 88888888888 ATLANTA GA',
           'BEDBATHBEYOND 898098098 snellville, GA',
            'ASHLEY FURNITURE 8****9999 ATHENS GA']}

# Putting Dict into a DF
---

In [34]:
temp_df = pd.DataFrame()
for key, value in transaction_dict.items():
    temp_df[key] = pd.Series(value)

In [35]:
#temp_df.head()

# Flipping DF
---

In [36]:
temp_df.columns = temp_df.columns.str.split("_", expand =True)
temp_df = temp_df.stack(level=0).rename_axis((None, 'identifier')).reset_index('identifier').reset_index(drop=True)
temp_df = temp_df.rename(columns={temp_df.columns[1]: 'transaction'})
temp_df.head()

identifier                                 transaction
0         gas           QuikTrip 0000 00000000 ATLANTA GA
1   groceries               KROGER #000 STONE MOUNTAIN GA
2    grooming  CRAFTED HAIR SALON JUo000 LAWRENCEVILLE GA
3  homeimprov                     LOWES #00000* BUFORD GA
4     medical               CVS/PHARMACY #0000 SUWANEE GA

# Adding Random Dates
---

In [37]:
def get_end_month(year, month):
    end = calendar.monthrange(year, month)
    return end[1] 

In [38]:
def generate_ran_date(year, month, first_day, end_day):
    start = datetime.date(year, month, first_day)
    end = datetime.date(year, month, end_day)
    time_btw = end - start
    days_btw = time_btw.days
    rand_num = random.randrange(days_btw)
    rand_date = start + datetime.timedelta(days=rand_num)
    return rand_date

In [39]:
dates = []
for i in range(len(temp_df)):
    dates.append(generate_ran_date(2018, 1, 1, 31))
temp_df['date'] = dates

### Making a List of Lists of the first and last day of the month for datetime interval

In [40]:
months = []
for i in range(13):
    if  i != 0:
        months.append([[2018, i, 1], [2018, i, get_end_month(2018, i)]])

In [41]:
temp_df.head()

identifier                                 transaction        date
0         gas           QuikTrip 0000 00000000 ATLANTA GA  2018-01-29
1   groceries               KROGER #000 STONE MOUNTAIN GA  2018-01-06
2    grooming  CRAFTED HAIR SALON JUo000 LAWRENCEVILLE GA  2018-01-20
3  homeimprov                     LOWES #00000* BUFORD GA  2018-01-14
4     medical               CVS/PHARMACY #0000 SUWANEE GA  2018-01-25

# Adding Transaction Amounts
---
- Making a sensible pattern to spending
- [US Bureau of Statistics Consumer Expendatures 2020](https://www.bls.gov/news.release/pdf/cesan.pdf)

In [42]:
temp_df['identifier'].unique()

array(['gas', 'groceries', 'grooming', 'homeimprov', 'medical',
       'conveyance', 'takeaway', 'utility', 'clothing', 'holiday',
       'streaming', 'work', 'gift', 'baby', 'wedding', 'furniture'],
      dtype=object)

### Average person drives 
- 14,263 miles/yr divided by 12 == 1189 miles/mo divided by 4 weeks == 297.25 miles/wk
- 297.25 miles/wk divided by 24 miles per gallon == 12.39
- Average cost of gas in 2018 == $2.81
- Monthly average spending
- [Historical Gas Prices](https://www.eia.gov/dnav/pet/hist/LeafHandler.ashx?n=pet&s=emm_epmr_pte_nus_dpg&f=m)

In [43]:
14263/24*2.813

1671.7424583333334

In [44]:
def gen_ran_num(start, stop):
    return random.randint(start, stop)
def gen_ran_float(start, stop):
    return random.uniform(start, stop)

# Adding gas purchases
- Using a random amount between the min price 11 and the max price 14 and a random amount up top 6
- Adding in gas purchases

In [45]:
gas_prices_by_mo = [2.671, 2.705, 2.709, 2.873, 2.987, 2.928, 2.928, 2.914, 2.915, 2.943, 2.736, 2.457]
gas = temp_df.loc[temp_df['identifier'] == 'gas'].reset_index(drop=True)
#multiplying jan gas prices by average gallons bought by week
gas['amount'] = [gas_prices_by_mo[0]*gen_ran_float(11, 14) for i in range(len(gas))]
gas

identifier                          transaction        date     amount
0        gas    QuikTrip 0000 00000000 ATLANTA GA  2018-01-29  33.012488
1        gas  SHELL OIL 00000000000 EAST POINT GA  2018-01-21  34.764647
2        gas          EXXON 000000-00 ATLANTA, GA  2018-01-04  36.760044
3        gas    QuikTrip 1234 00000000 ATLANTA GA  2018-01-12  33.258587
4        gas   SHELL OIL 1111000000 EAST POINT GA  2018-01-13  29.650713
5        gas          EXXON 222200-00 ATLANTA, GA  2018-01-24  36.245501

In [46]:
def create_random_gas(slice_df, max_duplicates, min_price, max_price, month, day, end_day):
    gas_prices_by_mo = [2.671, 2.705, 2.709, 2.873, 2.987, 2.928, 2.928, 2.914, 2.915, 2.943, 2.736, 2.457]
    random_purchases = gen_ran_num(1, max_duplicates)
    copy_df = pd.DataFrame(columns = slice_df.columns)
    if random_purchases <=len(slice_df):#####
        to_append = random_purchases
        for i in range(to_append):
            copy_df = copy_df.append(slice_df.iloc[gen_ran_num(0, len(slice_df)-1)]).reset_index(drop=True)
            copy_df.date[i] = generate_ran_date(2018, month, day, end_day)
            copy_df.amount[i] = gas_prices_by_mo[0]*gen_ran_float(min_price, max_price)
    else:
        for i in range(random_purchases):
            copy_df = copy_df.append(slice_df.iloc[gen_ran_num(0, len(slice_df)-1)]).reset_index(drop=True)
            copy_df.date[i] = generate_ran_date(2018, month, day, end_day)
            #multiplying the gallons by the gas price of the month purchased
            copy_df.amount[i] = gas_prices_by_mo[0]*gen_ran_float(min_price, max_price)
    return copy_df

In [47]:
# def create_random_gas(slice_df, max_duplicates, min_price, max_price, month, day, end_day):
#     gas_prices_by_mo = [2.671, 2.705, 2.709, 2.873, 2.987, 2.928, 2.928, 2.914, 2.915, 2.943, 2.736, 2.457]
#     random_purchases = gen_ran_num(1, max_duplicates)
#     #print(random_purchases)
#     #print(slice_df)
#     if random_purchases <=len(slice_df):#####
#         print('optin1')
#         to_append = random_purchases
#         slice_df = slice_df[random_purchases-1:]
#         for i in range(to_append):
#             slice_df = slice_df.append(slice_df.iloc[gen_ran_num(0, len(slice_df)-1)]).reset_index(drop=True)
#             slice_df.date[i] = generate_ran_date(2018, month, day, end_day)
#             slice_df.amount[i] = gas_prices_by_mo[0]*gen_ran_float(min_price, max_price)
#     else:
#         print('option2')
#         to_append = random_purchases - len(slice_df)
#         for i in range(to_append):
#             ####
#             slice_df = slice_df.append(slice_df.iloc[gen_ran_num(0, len(slice_df)-1)]).reset_index(drop=True)
#             slice_df.date[i] = generate_ran_date(2018, month, day, end_day)
#             #multiplying the gallons by the gas price of the month purchased
#             slice_df.amount[i] = gas_prices_by_mo[0]*gen_ran_float(min_price, max_price)
#     return slice_df

In [51]:
#create_random_gas(gas, 6, 11, 14, 2, 1, 28)

In [48]:
all_gas = pd.DataFrame()
for i in range(len(months)):
    if i > 0:
        all_gas = all_gas.append(create_random_gas(gas, 6, 11, 14, months[i][0][1], months[i][0][2], months[i][1][2])).reset_index(drop=True)
all_gas = all_gas.append(gas)

/Users/Jen/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':
/Users/Jen/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.


In [52]:
all_gas = all_gas.sort_values(by='date').reset_index(drop=True)

In [61]:
avg_tot_convey = keep.loc[keep['category'] == 'transportation'].reset_index(drop=True)['2018'][0]
(avg_tot_convey - all_gas.amount.sum())/12
#avg_tot_convey/12

677.1962048614038

In [56]:
transaction_dict['conveyance']

['E ZPASS MD CSC0000000 000-0000000 MD',
 'AUTOZONE 0000 STONEMOUNTAIN GA',
 'ADVANCE AUTO PARTS #0000 LITHONIA GA',
 'EXPRESSOILCHANGE 00*0000000-0 SNELLVILLE, GA',
 'NORTHERN TOOL 00 000000 000 SNELLVILLE, GA']

In [30]:
temp_df['identifier'].unique()

array(['gas', 'groceries', 'grooming', 'homeimprov', 'medical',
       'conveyance', 'takeaway', 'utility', 'clothing', 'holiday',
       'streaming', 'work', 'gift', 'baby', 'wedding', 'furniture'],
      dtype=object)

In [62]:
keep

category   2018   2019   2020  percentage_2020  \
0   average income before taxes  78635  82852  84352             0.00   
1   average annual expenditures  61224  63036  61334             0.00   
2                          food   7923   8169   7316            11.98   
3                       housing  20091  20679  21409            35.05   
4          apparel and services   1866   1883   1434             2.35   
5                transportation   9761  10742   9826            16.09   
6                    healthcare   4968   5193   5177             8.48   
7                 entertainment   3226   3050   2912             4.77   
8        personal_care_products    768    768    864             1.41   
9                       reading    108     92    114             0.19   
10                    education   1407   1443   1271             2.08   
11                      tobacco    347    320    315             0.52   
12                         misc    993    899    907             1.49   
13                 cash_contrib   1888   1995   2283             3.74   
14  personal_insurance_pensions   7296   7165   7246            11.86   

    percentage_2019  percentage_2018  
0              0.00             0.00  
1              0.00             0.00  
2             13.09            13.07  
3             33.14            33.13  
4              3.02             3.08  
5             17.22            16.10  
6              8.32             8.19  
7              4.89             5.32  
8              1.23             1.27  
9              0.15             0.18  
10             2.31             2.32  
11             0.51             0.57  
12             1.44             1.64  
13             3.20             3.11  
14            11.48            12.03

In [32]:
# Average amount spent on groceries monthly
keep.loc[keep['category'] == 'food'].reset_index(drop=True)['2018'][0]/12

660.25

# Description of person
- Average income before taxes 90,000
- Has 80,000 in savings
- Male, single, dating, June - December 2017
- Renting 2500/mo 33% of take home
- January 2018 25% eating out, 75% groceries
- January 2018 buys ring, 4,000
- January 2018 engaged
- March 2018 buys tux
- April 2018 Files as single
- April 2018 gets wedding planner
- June 2018 married
- July 2018 honeymoon
- April 2019 Files as Married Joint
- January 2019 buys house
- March - November buys baby stuff
- November Baby 2019


- Watches online streaming
- Goes to movies 1X mo
- buys winter clothes christmas December
- buys summer clothes birthday May
